src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs

```diff
@@ -0,0 +1,133 @@
+    private readonly ISmartRenameSession _smartRenameSession;
+#pragma warning restore CS0618 
+
+    private readonly IThreadingContext _threadingContext;
+
+    private readonly CancellationTokenSource _cancellationTokenSource = new();
+
+    public event PropertyChangedEventHandler? PropertyChanged;
+
+    public ObservableCollection<string> SuggestedNames { get; } = new ObservableCollection<string>();
+
+    public bool IsAvailable => _smartRenameSession?.IsAvailable ?? false;
+
+    public bool HasSuggestions => _smartRenameSession?.HasSuggestions ?? false;
+
+    public bool IsInProgress => _smartRenameSession?.IsInProgress ?? false;
+
+    public string StatusMessage => _smartRenameSession?.StatusMessage ?? string.Empty;
+
+    public bool StatusMessageVisibility => _smartRenameSession?.StatusMessageVisibility ?? false;
+
+    private string? _selectedSuggestedName;
+
+    /// <summary>
+    /// The last selected name when user click one of the suggestions. <see langword="null"/> if user hasn't clicked any suggestions.
+    /// It would trigger <see cref="PropertyChanged"/> when it get changed.
+    /// </summary>
+    public string? SelectedSuggestedName
+    {
+        get => _selectedSuggestedName;
+        set
+        {
+            if (_selectedSuggestedName != value)
+            {
+                _threadingContext.ThrowIfNotOnUIThread();
+                _selectedSuggestedName = value;
+                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(SelectedSuggestedName)));
+            }
+        }
+    }
+
+    public SmartRenameViewModel(
+        IThreadingContext threadingContext,
+        IAsynchronousOperationListenerProvider listenerProvider,
+#pragma warning disable CS0618 // Editor team use Obsolete attribute to mark potential changing API
+        ISmartRenameSession smartRenameSession)
+#pragma warning restore CS0618
+    {
+        _threadingContext = threadingContext;
+        _smartRenameSession = smartRenameSession;
+        _smartRenameSession.PropertyChanged += SessionPropertyChanged;
+        var listener = listenerProvider.GetListener(FeatureAttribute.SmartRename);
+
+        using var listenerToken = listener.BeginAsyncOperation(nameof(_smartRenameSession.GetSuggestionsAsync));
+        _smartRenameSession.GetSuggestionsAsync(_cancellationTokenSource.Token);
+    }
+
+    private void SessionPropertyChanged(object sender, PropertyChangedEventArgs e)
+    {
+        _threadingContext.ThrowIfNotOnUIThread();
+        // _smartRenameSession.SuggestedNames is a normal list. We need to convert it to ObservableCollection to bind to UI Element.
+        if (e.PropertyName == nameof(_smartRenameSession.SuggestedNames))
+        {
+            SuggestedNames.Clear();
+            foreach (var name in _smartRenameSession.SuggestedNames)
+            {
+                SuggestedNames.Add(name);
+            }
```

Comment:
i think an observable collection has a way to say "i'm about to make a bunch of changes" so it doesn't notify on each change.